matlab/HyEQsolver_new_slow.m

```diff
@@ -0,0 +1,434 @@
+        
+        if length(t_flow) == 1
+            % Handle the case where the ode solver doesn't step forward in time.
+            % This is ony known to happen when the solution is blowing up to
+            % infinity. 
+            warning(horzcat('The hybrid solver is not progressing at t=%.4g, j=%d, ', ...
+                'possibly due to |f(x, t, j)|=%0.3g being very large.\n', ...
+                'Aborting solver and setting final value of x to NaN.'), ...
+                t_flow, j, norm(f(x_flow', t_flow, j)))
+            
+            % We check if we have preallocated enough memory
+            if (index_last + 1 > size(t_total, 1))
+                x_total = [x_total; zeros(1, cx)];
+                t_total = [t_total; 0];
+                j_total = [j_total; 0];
+            end     
+
+            x_total(index_last + 1, :) = NaN(size(x_flow));
+            t_total(index_last + 1, :) = t_flow;
+            j_total(index_last + 1, :) = j;
+            index_last = index_last + 1;
+            
+            break
+        end
+        
+        % Matlab ODE solvers miss events if they occur at the second time
+        % step (such as the case as if the state in the boundary of a
+        % closed set). To correct for this, we manually check if the second
+        % state entry is at a point where flowing is allowed. If it is,
+        % then we simply append the solution from the ODE solver to the
+        % output. But if it's not, then we try time steps from the initial
+        % point to find the longest time step that takes us out of C (or
+        % into D, if using jump priority). The next state value is then
+        % calculated using linear interpolation and appended to the output.
+        second_state_in_C = C(x_flow(2, :)',t_flow(2),j);
+        second_state_in_D = D(x_flow(2, :)',t_flow(2),j);
+        will_second_state_flow = second_state_in_C && ~(second_state_in_D && rule == 1);
+        if will_second_state_flow || length(t_flow) == 2
+            nt = length(t_flow);
+            
+            % We add the new interval of flow to the solution--omitting the
+            % first entry because it duplicates the previous entry in the
+            % solution.
+            
+            % We check if we have preallocated enough memory, and if not,
+            % we add additional entries.
+            while (index_last + nt - 1 > size(t_total, 1))
+                x_total = [x_total; zeros(prealloc_size, cx)];
+                t_total = [t_total; zeros(prealloc_size, 1)];
+                j_total = [j_total; zeros(prealloc_size, 1)];
+            end
+            
+            x_total(index_last + 1 : index_last + nt - 1, :) = x_flow(2:end,:); 
+            t_total(index_last + 1 : index_last + nt - 1, :) = t_flow(2:end); 
+            j_total(index_last + 1 : index_last + nt - 1, :) = j*ones(nt-1, 1); 
```

Comment:
The expression `index_last + nt - 1` is used several times. Maybe define a variable for it, or perhaps just for `nt - 1`.